### PR TITLE
change: proxy-rewrite plugin add request headers rewriting

### LINKED ISSUE #628: change: proxy-rewrite plugin add request headers rewriting
Added request header rewrite feature.

增加请求头重写功能。

### PR DESCRIPTION
**Pull request by shuaijinchao**
### Issues resolved

Fix #628

### CODE REVIEW

File: t/plugin/proxy-rewrite.t [hunk 1]
```diff
@@ -403,3 +403,61 @@ host: localhost
...
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "plugins": {
+                            "proxy-rewrite": {
+                                "uri": "/uri/plugin_proxy_rewrite",
+                                "headers": {
+                                    "X-Api-Version": "v2",
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: we should add more test cases for exceptions.

File: doc/plugins/proxy-rewrite-cn.md [hunk 2]
```diff
@@ -10,6 +10,7 @@
 |uri             |可选| 转发到上游的新`uri` 地址|
 |host            |可选| 转发到上游的新`host` 地址，例如：`iresty.com` |
 |enable_websocket|可选| 是否启用`websocket`（布尔值），默认不启用|
+|headers         |可选| 转发到上游的新`headers`，可以设置多个。头信息如果存在将重写，不存在则添加。|
```
[the remark this answers is not in the document]
Comment (shuaijinchao) [reply doc/plugins/proxy-rewrite-cn.md]: OK, There is no deletion function at the moment. I need to add it.

[the remark this answers is not in the document]
Comment (membphis) [reply doc/plugins/proxy-rewrite-cn.md]: Can we use "null" to indicate the delete action?

[the remark this answers is not in the document]
Comment (shuaijinchao) [reply doc/plugins/proxy-rewrite-cn.md]: OK. It's more reasonable to use "null".

File: lua/apisix/core/request.lua [hunk 3]
```diff
@@ -19,6 +22,27 @@ local function _headers(ctx)
...
 end
+
+local function _validate_header_name(name)
+    local tname = type(name)
+    if tname ~= "string" then
+        return nil, str_fmt("invalid header name %q: got %s, " ..
+                "expected string", name, tname)
+    end
+
+    return name, nil
```
Comment (moonming) [inline lua/apisix/core/request.lua]: `return name` is enough.

File: lua/apisix/core/request.lua [hunk 4]
```diff
@@ -19,6 +22,27 @@ local function _headers(ctx)
...
+end
+
+local function _validate_header_value(value)
+    local tvalue = type(value)
+    if tvalue ~= "string" and tvalue ~= "number" and tvalue ~= "boolean" then
+        return nil, str_fmt("invalid header value for %q: got %s, expected " ..
+                    "string, number or boolean", value, tvalue)
+    end
+
+    return value, nil
```
Comment (moonming) [inline lua/apisix/core/request.lua]: ditto

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 5]
```diff
@@ -74,6 +79,17 @@ function _M.rewrite(conf, ctx)
         ctx.var["upstream_upgrade"]    = ctx.var["http_upgrade"]
         ctx.var["upstream_connection"] = ctx.var["http_connection"]
     end
+
+    if conf.headers then
+        for header_name, header_value in pairs(conf.headers) do
+            if header_value == "null" then
+                -- Remove Header
+                core.request.set_header(header_name, "")
```
Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: should be `core.request.set_header(header_name, nil)`? I am not sure.

File: lua/apisix/core/request.lua (hunk 3 above)
Comment (shuaijinchao) [reply lua/apisix/core/request.lua]: return only the `name` or cause the use case `t/node/invalid-service.t` to throw an exception.

File: lua/apisix/core/request.lua (hunk 4 above)
Comment (shuaijinchao) [reply lua/apisix/core/request.lua]: ditto

File: lua/apisix/plugins/proxy-rewrite.lua (hunk 5 above)
Comment (shuaijinchao) [reply lua/apisix/plugins/proxy-rewrite.lua]: currently `core.request.set_header` checks the parameter of `header value`, only `string`, `number`, `boolean` types are allowed. Other types will throw an exception.

File: doc/plugins/proxy-rewrite.md [hunk 6]
```diff
@@ -26,7 +27,12 @@ curl http://127.0.0.1:9080/apisix/admin/routes/1 -X PUT -d '
             "uri": "/test/home.html",
             "scheme": "http",
             "host": "iresty.com",
-            "enable_websocket": true
+            "enable_websocket": true,
+            "headers": {
+                "X-Api-Version": "v1",
+                "X-Api-Engine": "apisix",
+                "X-Api-Test": "null"
```
Comment (moonming) [inline doc/plugins/proxy-rewrite.md]: the `"null"` is not good, if the string value is `"null"`, how to deal with it?
it should be `null`.

File: lua/apisix/plugins/proxy-rewrite.lua [hunk 7]
```diff
@@ -74,6 +79,17 @@ function _M.rewrite(conf, ctx)
         ctx.var["upstream_upgrade"]    = ctx.var["http_upgrade"]
         ctx.var["upstream_connection"] = ctx.var["http_connection"]
     end
+
+    if conf.headers then
+        for header_name, header_value in pairs(conf.headers) do
+            if header_value == "null" then
```
Comment (moonming) [inline lua/apisix/plugins/proxy-rewrite.lua]: we can not use string `"null"` here.

File: t/plugin/proxy-rewrite.t [hunk 8]
```diff
@@ -403,3 +403,223 @@ host: localhost
...
+
+=== TEST 15: rewrite headers
+--- request
+GET /hello HTTP/1.1
+--- more_headers
+X-Api-Version:v1
+--- response_body
+uri: /uri/plugin_proxy_rewrite
+host: localhost
+x-api-version: v2
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: why not use `response-header` to judge headers?

File: lua/apisix/core/request.lua (hunk 3 above)
Comment (moonming) [reply lua/apisix/core/request.lua]: I think it's nonbusiness with `return name`. I will try to dig.

File: lua/apisix/core/request.lua [hunk 9]
```diff
@@ -19,6 +22,27 @@ local function _headers(ctx)
...
+        return nil, str_fmt("invalid header name %q: got %s, " ..
+                "expected string", name, tname)
+    end
+
+    return name, nil
+end
+
+local function _validate_header_value(value)
+    local tvalue = type(value)
+    if tvalue ~= "string" and tvalue ~= "number" and tvalue ~= "boolean" then
```
Comment (moonming) [inline lua/apisix/core/request.lua]: I think we can remove the type check

File: lua/apisix/core/request.lua (hunk 3 above)
Comment (shuaijinchao) [reply lua/apisix/core/request.lua]: Yes, it's not related to the name of the return parameter. What I want to express is related to the number of parameters.

File: doc/plugins/proxy-rewrite.md (hunk 6 above)
Comment (shuaijinchao) [reply doc/plugins/proxy-rewrite.md]: empty string is good, l revise it.

File: lua/apisix/core/request.lua (hunk 9 above)
Comment (shuaijinchao) [reply lua/apisix/core/request.lua]: Why?

File: doc/plugins/proxy-rewrite.md (hunk 6 above)
Comment (membphis) [reply doc/plugins/proxy-rewrite.md]: $ resty -e 'print(require("cjson").encode({a=ngx.null}))'
{"a":null}

how about we use `null` ?

File: lua/apisix/core/request.lua (hunk 9 above)
Comment (moonming) [reply lua/apisix/core/request.lua]: The header_value can take an array list of values, for example,
```
 ngx.req.set_header("Foo", {"a", "abc"})
```
will produce two new request headers:
```
 Foo: a
 Foo: abc
```

File: doc/plugins/proxy-rewrite.md (hunk 6 above)
Comment (shuaijinchao) [reply doc/plugins/proxy-rewrite.md]: @moonming What about you?

Comment (shuaijinchao) [reply doc/plugins/proxy-rewrite.md]: I think empty strings are better.

Comment (moonming) [reply doc/plugins/proxy-rewrite.md]: I think we should keep the same style with OpenResty API:
```
 ngx.req.set_header("X-Foo", nil)
```
And I am not sure we should use `null` or `cjson.null` here, you can make a test.

File: lua/apisix/core/request.lua (hunk 3 above)
Comment (moonming) [reply lua/apisix/core/request.lua]: it's run ok on my side after remove useless `nil`.

File: t/plugin/proxy-rewrite.t [hunk 10]
```diff
@@ -403,3 +403,223 @@ host: localhost
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 19: rewrite empty headers
```
Comment (moonming) [inline t/plugin/proxy-rewrite.t]: should be `TEST 21`.
Please use https://github.com/openresty/openresty-devel-utils/blob/master/reindex to format test cases.

File: t/plugin/proxy-rewrite.t (hunk 8 above)
Comment (membphis) [reply t/plugin/proxy-rewrite.t]: we used a fake server to response the request header for checking @moonming

File: lua/apisix/core/request.lua (hunk 3 above)
Comment (shuaijinchao) [reply lua/apisix/core/request.lua]: fixed.

**Comment by moonming** [comment]
@shuaijinchao please add doc and test cases for `null` as the value of header. else looks good to me.
